arch/xtensa/core/mpu.c

```diff
@@ -0,0 +1,1235 @@
+/*
+ * Copyright (c) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <string.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/toolchain.h>
+#include <zephyr/arch/xtensa/arch_inlines.h>
+#include <zephyr/arch/xtensa/mpu.h>
+#include <zephyr/linker/linker-defs.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util_macro.h>
+
+#include <xtensa/corebits.h>
+#include <xtensa/config/core-matmap.h>
+#include <xtensa/config/core-isa.h>
+#include <xtensa_mpu_priv.h>
+
+extern char _heap_end[];
+extern char _heap_start[];
+extern char __data_start[];
+extern char __data_end[];
+
+#ifndef CONFIG_XTENSA_MPU_ONLY_SOC_RANGES
+__weak const struct xtensa_mpu_range xtensa_soc_mpu_ranges[0];
+__weak int xtensa_soc_mpu_ranges_num;
+#endif
+
+/** MPU foreground map for kernel mode. */
+static struct xtensa_mpu_map xtensa_mpu_map_fg_kernel;
+
+/*
+ * Additional information about the MPU maps: foreground and background
+ * maps.
+ *
+ *
+ * Some things to keep in mind:
+ * - Each MPU region is described by TWO entries:
+ *   [entry_a_address, entry_b_address). For contiguous memory regions,
+ *   this should not much of an issue. However, disjoint memory regions
+ *   "waste" another entry to describe the end of those regions.
+ *   We might run out of available entries in the MPU map because of
+ *   this.
+ *   - The last entry is a special case as there is no more "next"
+ *     entry in the map. In this case, the end of memory is
+ *     the implicit boundary. In another word, the last entry
+ *     describes the region between the start address of this entry
+ *     and the end of memory.
+ * - Current implementation has following limitations:
+ *   - All enabled entries are grouped towards the end of the map.
+ *   - No disabled MPU entries allowed in between.
+ *   - Background map entries are copied to foreground map to
+ *     ensure correct alignment for hardware... which means it may
+ *     waste some foreground entries if there are no custom regions
+ *     between background entries.
```

Comment:
FWIW: the "background" feature in the hardware doesn't seem very useful to me.  Existing hardware I can see core-isa.h files for never does anything more complicated than e.g. "all of memory is mapped uncached".  I'd suggest dropping this code and just using the pre-existing SOC-provided memory map we already have.  At most, we should just be validating the map provided by the background vs. what we know from dts/kconfig to be "The Truth" as a check on SOC config.

Reply:
AFAIK, if you cross a background map "boundary", the address needs to be appear in the foreground map. In other words, any foreground regions cannot cross the boundaries of any background regions. And also that the entry[0], if enabled, must be the same as one of the background entries. So I just take the easy and simpler route here of copying the BG entries to FG as starting point. Having to account for BG entries when add new regions is going to complicate the code quite a bit... as the logic now is already overly complicated, I simply do not want to add to the confusion.

Reply:
To be clear: my suggestion is "ignore the background mapping" and just have a fully-populated mapping in the foreground (which takes no more entries).  The requirement of not crossing background boundaries regions is AFAICT incorrect, and if it's not a requirement there's little to no value in trying to support the background stuff.

The only reason we'd even want it would be if you had a device with genuinely useful fine-grained background mappings that would allow us to save foreground entries for register blocks or whatever.  But that seems not to be the case on the hardware I've seen where "background" just means "something that works OK at boot" (e.g. on mt8195 it's all uncached, which is basically useless).

Reply:
Do you happen to have simulator or hardware setup to test this quickly? I can try it on QEMU and see.